### ISSUE TITLE
remote websocket connection closed: Cannot write to closing transport

### ISSUE THREAD

**Issue by el-Presi** [ReneNulschDE/mbapi2020#279]
### The problem

Since a few days now (and I noticed a slightly design change within the Mercedes ME app also in that time), I cannot send commands to the car anymore (remote websocket connection closed: Cannot write to closing transport).
When I perform a restart of the integration, it works again

### What version of MBAPI2020 do use?

v0.19.1

### What type of installation are you running?

Home Assistant OS

### Anything in the logs that might be useful for us?

```txt
Dieser Fehler stammt von einer benutzerdefinierten Integration

Logger: custom_components.mbapi2020.websocket
Quelle: custom_components/mbapi2020/websocket.py:229
Integration: MercedesME 2020 (Dokumentation, Probleme)
Erstmals aufgetreten: 17. November 2024 um 21:15:27 (1 Vorkommnisse)
Zuletzt protokolliert: 17. November 2024 um 21:15:27

Other error 429, message='Invalid response status', url='wss://websocket.emea-prod.mobilesdk.mercedes-benz.com/v2/ws'


- - - - - - - - - - - - - - - - -


Dieser Fehler stammt von einer benutzerdefinierten Integration

Logger: custom_components.mbapi2020.client
Quelle: custom_components/mbapi2020/client.py:263
Integration: MercedesME 2020 (Dokumentation, Probleme)
Erstmals aufgetreten: 17. November 2024 um 21:15:27 (1 Vorkommnisse)
Zuletzt protokolliert: 17. November 2024 um 21:15:27

Error with the websocket connection (retry counter: 0):


- - - - - - - - - - - - - -


Dieser Fehler stammt von einer benutzerdefinierten Integration

Logger: custom_components.mbapi2020.websocket
Quelle: custom_components/mbapi2020/websocket.py:179
Integration: MercedesME 2020 (Dokumentation, Probleme)
Erstmals aufgetreten: 10:37:30 (4 Vorkommnisse)
Zuletzt protokolliert: 10:42:39

remote websocket connection closed: Cannot write to closing transport
```

**Comment by ReneNulschDE** [comment]
I see this error since last Tuesday. MB has changed some settings in their backend and closing the sessions more often. (every 15min now) and http429 is a "Too many requests block" - have not found a solution yet.

Could you enable the debug logging, wait for the error and share the complete log (via email to mbdev.i279@nulsch.de)?

**Comment by ReneNulschDE** [comment]
The MB-Team changed the server behavior this morning to the old setting from last week. So this problem should not happen anymore. I'll close this but feel free to open it again with detailed logs.

**Comment by dennzs** [comment]
The MB app was down for a few hours last night so I wonder if perhaps they broke something and then reverted a change because of that.

**Comment by ethemcemozkan** [comment]
I started experienced the same error today as well. Perhaps they reverted their revert? :) When I have debug logs I will share with you

**Comment by iHaveAstream** [comment]
Hi, jist fyi: the issue is still present.
Got the very same error also this morning when the integration tried to enable the auxheater via HA automation.

**Comment by el-Presi** [comment]
Just FYI: I am still waiting for the error to happen again, then I will be able to post some logs :)

**Comment by AlirezaT** [comment]
It's a few days that after restarting HA the integration doesn't work:
`2024-11-20 21:07:33.423 ERROR (MainThread) [custom_components.mbapi2020.websocket] Other error 429, message='Invalid response status', url='wss://websocket.emea-prod.mobilesdk.mercedes-benz.com/v2/ws'
2024-11-20 21:07:33.423 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 0): 
2024-11-20 21:08:03.904 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry xxxxx@gmail.com (Region: Europe) for mbapi2020
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 635, in __async_setup_with_context
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/mbapi2020/__init__.py", line 207, in async_setup_entry
    raise HomeAssistantError(
homeassistant.exceptions.HomeAssistantError: No car information registered for this account. Check the MB website with the same account. Giving up...`

**Comment by dennzs** [comment]
Mine also stopped working again after a few days but the app is also acting up and unable to synchronize. Not just the integration to HA.

**Comment by AlirezaT** [comment]
Indeed, I just checked the app and it's not syncing. However, I can log in to the website!

**Comment by iHaveAstream** [comment]
I can control the car via the app.
And with the integration it works when I reload the integration. 
However, I think tomorrow morning it would not work anymore before reloading the integration.

**Comment by UDicke** [comment]
I just reloaded 2 hubs and they delivered data again. I will report when working stops again...

**Comment by paludi** [comment]
I can confirm that it works again for me as well. I added my new car yesterday which failed with the same errors. (even in the Me app) 
Since today morning HA integration and Me app seem to be working again.

**Comment by iHaveAstream** [comment]
Hi, also in my case it's working at the moment and the integration was able to activate the auxheater via automation this morning, as usual.
Looks promising.

**Comment by ReneNulschDE** [comment]
It is not solved my dev and test environment stopped again. I'm still collecting information.

For further answers, please include how many accounts you have integrated and the count of devices that each account is using. (Example: 1 Account, used in two HA-Env and 1 Phone, 2 Pads, 1 Watch)

**Comment by dannyfiorot** [comment]
Same error, but try this:

Enter in the MERCEDES ME INTEGRATION, re-add your car using the email that you use to log-in in the Mercedes ME app and then add it without cancel the not working entities.

**Comment by iHaveAstream** [comment]
> Same error, but try this:
> 
> Enter in the MERCEDES ME INTEGRATION, re-add your car using the email that you use to log-in in the Mercedes ME Application.
> 
> The error in my Home Assistant confuguration is created by using the VIN and PIN information in the setup steps. If you install the MERCEDES ME integration using only the email, all work fine.

Doesn't seem to work for me.
In the web portal (https://mercedes.me) I use my mobile number to access it.
However, the integration seems not to recognise my mobile number, even though I entered it correctly incl. leading country code (+49).

When I try instead with my registered e-mail address,  I get:

```
Logger: homeassistant.config_entries
Source: config_entries.py:635
First occurred: 21:46:01 (1 occurrences)
Last logged: 21:46:01

Error setting up entry **********@gmail.com (Region: Europe) for mbapi2020
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 635, in __async_setup_with_context
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/mbapi2020/__init__.py", line 207, in async_setup_entry
    raise HomeAssistantError(
homeassistant.exceptions.HomeAssistantError: No car information registered for this account. Check the MB website with the same account. Giving up...
```

**Comment by weesteev** [comment]
Having the same issue today, have removed integration and re-added but just fails now. Also noticed the app seems to be showing incorrect data and still thinks my car is charging when it was last plugged in about 10 hours ago. Think this is definitely a back end issue with Mercedes.
![Screenshot 2024-11-21 222035](https://github.com/user-attachments/assets/6bf7bc6a-41e7-4b60-9a03-226108a1c6d1)

**Comment by silashansen** [comment]
> Having the same issue today, have removed integration and re-added but just fails now. Also noticed the app seems to be showing incorrect data and still thinks my car is charging when it was last plugged in about 10 hours ago. Think this is definitely a back end issue with Mercedes. Image below of my error log from HA.
> 
> ![Screenshot 2024-11-21 222035](https://private-user-images.githubusercontent.com/76874416/388718384-6bf7bc6a-41e7-4b60-9a03-226108a1c6d1.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIyNjAzNzYsIm5iZiI6MTczMjI2MDA3NiwicGF0aCI6Ii83Njg3NDQxNi8zODg3MTgzODQtNmJmN2JjNmEtNDFlNy00YjYwLTlhMDMtMjI2MTA4YTFjNmQxLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIyVDA3MjExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3MjA1NjMwMGNlZTU5MGI1MjU4MDM0ZWU1ZDdlY2E4ODNkMWU3MTAzODAxMDY3MTk3YTFhZWE1ZmViYTU4NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8QYc6uDogbJ-Ex_sa50hbZZRDcoMBNquJdDmPjKCoYc)

I've experienced it a few times as well, but reloading the integration has been fixing it every time. I think you just ran into trying to re-auth while the issue was ongoing. It will probably work if you try it again now.

Regarding the issue:
It's fair that the Mercedes backend is failing or returning 429 from time to time, but we should just handle this more gracefully, to avoid leaving the integration in a permanently failed state after the error occurs. I'm sure it would be trivial for @ReneNulschDE to wrap this in some more graceful error handling. I'd love to do this, but I barely have a dev environment up and running for this integration at this point (will work on getting setup).

**Comment by iHaveAstream** [comment]
Thanks for the hint @silashansen!
I was able to re-auth the integration just this moment (via mobile number).

**Comment by ReneNulschDE** [comment]
> Regarding the issue:
> It's fair that the Mercedes backend is failing or returning 429 from time to time, but we should just handle this more gracefully, to avoid leaving the integration in a permanently failed state after the error occurs. I'm sure it would be trivial

I just do not know how long we have to wait. in my tests 60min was too short and 4 hours too long. I'm still testing... - I made another change to force a reconnect from the client side after 12 min (the server breaks the connection after 15 min) to act more like a phone user. (I tested it with a reconnect every 30sec for 24h without a http429)

**Comment by ReneNulschDE** [comment]
Would be great if someone can test this beta release:
https://github.com/ReneNulschDE/mbapi2020/releases/tag/v0.19.2-beta.1

**Comment by ReneNulschDE** [comment]
This will not solve it - just got a 429 again...

**Comment by iHaveAstream** [comment]
Just fyi: for me, it still works ever since. At least it seems like. And this since 12 hours now.

**Comment by dannyfiorot** [comment]
Situation gone bad... Nothing working.... Integration cancelled and reinstalled....

**Comment by dcigir** [comment]
It doesn't work for me either. Even the Mercedes app is not showing info from the car :(

**Comment by dannyfiorot** [comment]
In 2025 all the car with 3g Module don't work either.... I'm disappointed.....

**Comment by willembuys** [comment]
I currently cannot use the app (1 phone, app has old data) or the integration (429 error) with my account (2 cars), but when I use the app on my partner's phone (with my partner's account, same 2 cars) while not on Wi-Fi, I can use the app.
So is it linked to how many connections are made with an account? And does it then also block the IP address?

**Comment by ReneNulschDE** [comment]
It's per account and not per IP based on my tests.

After 3-6 hours it will come back.

**Comment by ReneNulschDE** [comment]
I have published a new beta release. This version contains a new reconnect logic with an increasing wait time.

**Comment by StefanHesse** [comment]
I just installed the v0.19.2-beta.2 and have the automation running that i will be informed whenever the 429 error is appearing. I will give feedback as soon as the error appears.
1 Account, used in 1 HA, 1 iPAD, 1 Android phone

**Comment by ReneNulschDE** [comment]
I have deleted the beta-2 - and will republish a new version later today.

**Comment by StefanHesse** [comment]
OK, do you recommend me to go back to another version until then?

**Comment by willembuys** [comment]
Do you have an idea how much requests are allowed and how to limit the number of requests?

**Comment by ReneNulschDE** [comment]
I assume that it is not related to requests. Based on my tests its related to active app-ids per account/day. A new app-id is created on restart of HA, reload of the integration or connection problems.

I can't reuse the app-id with the current architecture of the component as the MB-Servers will not deliver a full-update and I create the sensors based on this first full-update.

**Comment by dennzs** [comment]
@ReneNulschDE when you say account do you mean a user or a car? Would it make sense to create a second (well third) account for our family and use that only with HA and keep the original for the MB app? 

I’ll check if my wife’s account works in the app the next time mine dies.

**Comment by ReneNulschDE** [comment]
I have released v0.19.2-beta.3 with an additional fix of the app-id handling.

**Comment by paludi** [comment]
@ReneNulschDE thanks a lot for your work and quick response!
I updated to beta3 and now communication with the Me servers work again. Before it was broken after upgrading to beta2 and reverting to 0.19.1.

I have 1 account with 1 HA instance and 1 android phone.

**Comment by ethemcemozkan** [comment]
If it matters, I started noticing frequent logouts from all MB apps(me, finance, eco...).
My setup: 2 accounts, one purely using the MB me app, the other one used for MB me, HA, and other companion apps

**Comment by firstcolle** [comment]
Beta 1 was fine for me. Installed beta 3 and I cant see my entityes. The integration inizialises fine. 
I noticed also that the data in the mb app is not updated, but is uodated in the web site

**Comment by ReneNulschDE** [comment]
> I cant see my entityes. The integration inizialises fine.

In beta 3, I only addressed the reconnection. When you restart HA and the account is blocked then the sensor get not initialized and connection restore on midnight will not initialize the sensors again. So you need to reload manually.

I'm working on beta 4, here I'll try to implement a reload of the component after the account got unblocked at midnight.

... midnight in the GMT timezone.

**Comment by Hooshee** [comment]
Maybe this helps a bit: 
2024-11-22 21:00:45.772 INFO (MainThread) [custom_components.mbapi2020.websocket] Connecting to wss://websocket.emea-prod.mobilesdk.mercedes-benz.com/v2/ws
2024-11-22 21:00:45.870 ERROR (MainThread) [custom_components.mbapi2020.websocket] Other error 429, message='Invalid response status', url='wss://websocket.emea-prod.mobilesdk.mercedes-benz.com/v2/ws'
2024-11-22 21:00:45.871 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 0): 
2024-11-22 21:00:45.871 INFO (MainThread) [custom_components.mbapi2020.client] Client WS Handler loop - loop end round 1
2024-11-22 21:01:00.872 INFO (MainThread) [custom_components.mbapi2020.client] Client WS Handler loop - loop end round 1
2024-11-22 21:01:15.873 INFO (MainThread) [custom_components.mbapi2020.client] Client WS Handler loop - loop end round 1
2024-11-22 21:01:30.875 INFO (MainThread) [custom_components.mbapi2020.client] Client WS Handler loop - loop end round 1
2024-11-22 21:01:45.876 INFO (MainThread) [custom_components.mbapi2020.client] Client WS Handler loop - loop end round 1
2024-11-22 21:02:00.877 INFO (MainThread) [custom_components.mbapi2020.client] Client WS Handler loop - loop end round 1
2024-11-22 21:02:15.878 INFO (MainThread) [custom_components.mbapi2020.client] Client WS Handler loop - loop end round 1
2024-11-22 21:02:30.879 INFO (MainThread) [custom_components.mbapi2020.client] Client WS Handler loop - loop end round 1
2024-11-22 21:02:42.110 DEBUG (MainThread) [custom_components.mbapi2020] starting lock
2024-11-22 21:02:42.110 INFO (MainThread) [custom_components.mbapi2020.client] Start Doors_lock for vin WDD21XXXXXXXX0706
2024-11-22 21:02:42.111 ERROR (MainThread) [custom_components.mbapi2020.websocket] remote websocket connection closed: Cannot write to closing transport

So, I arrived home, the location was changed to home, and then after 10 minutes the automation would close the car, if I leave it open. And since then it's not working, so the issue may start when you want to control something, not just read.

**Comment by ReneNulschDE** [comment]
Here is a [link](https://community.home-assistant.io/t/mercedes-me-component/41911/2199) to the current release notes of the upcoming version. I just need to wait until blocking and unblocking happens this to check if my test accounts work as expected...

**Comment by dennzs** [comment]
I may be daft here but how does one change the account used for the integration? Delete? Rerun the config somehow?

**Comment by ReneNulschDE** [comment]
Click on the Configure link on the integration page. There is an option to delete the auth information.

**Comment by ReneNulschDE** [comment]
v0.20.0 is published.

**Comment by dennzs** [comment]
Interestingly I was not able to fully finalize a new account yesterday while my main account was locked. Are they doing some IP based blocking as well or was this a separate issue? 

I could create the account but not use it to log in. This morning logging in with the new account worked and I could connect it to my car and HA successfully.

**Comment by iHaveAstream** [comment]
Hi,

This is what I got this night. I'm using a dedicated account (secondary) in the MB app now and the primary account for HA only.
Nevertheless, everything used to work this morning, so those logs can be ignored?

```
[custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 0): 429 - Invalid response status
2024-11-25 21:07:14.650 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 1): 429 - Invalid response status
2024-11-25 21:08:14.731 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 2): 429 - Invalid response status
2024-11-25 21:10:29.806 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 3): 429 - Invalid response status
2024-11-25 21:14:29.916 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 4): 429 - Invalid response status
2024-11-25 21:20:45.324 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 5): 429 - Invalid response status
2024-11-25 21:29:45.599 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 6): 429 - Invalid response status
2024-11-25 21:42:00.798 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 7): 429 - Invalid response status
2024-11-25 21:58:01.587 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 8): 429 - Invalid response status
2024-11-25 22:18:16.825 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 9): 429 - Invalid response status
2024-11-25 22:43:16.978 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 10): 429 - Invalid response status
2024-11-25 23:13:32.098 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 11): 429 - Invalid response status
2024-11-25 23:49:32.306 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 12): 429 - Invalid response status
2024-11-26 00:00:28.185 ERROR (MainThread) [aioshelly.rpc_device.wsrpc] Invalid Message from host 10.10.30.37:80: Received message error
2024-11-26 00:31:47.601 ERROR (MainThread) [custom_components.mbapi2020.client] Error with the websocket connection (retry counter: 13): 429 - Invalid response status
2024-11-26 01:37:48.632 ERROR (MainThread) [custom_components.mbapi2020.websocket] Could not connect: Cannot connect to host websocket.emea-prod.mobilesdk.mercedes-benz.com:443 ssl:default [Connect call failed ('20.67.76.32', 443)], retry in 10 seconds...
```

**Comment by ReneNulschDE** [comment]
@iHaveAstream : This means you got blocked at 21:07 and around 01:37 the connection was restored and the component reloaded. (the component update does what it should to do) - Lets see what happens tonight when you not restart HA or reload the component.

**Comment by ReneNulschDE** [comment]
@dennzs: I have not seen an IP based blocking. My dev and test environment is running on the same box. The dev account got blocked around 3pm and the test account at 9pm. We have to monitor it...

**Comment by dennzs** [comment]
@Amokd0c you shouldn't link the new account as if you were linking a new car. 

1. Instead, while in your MAIN account, go to the hamburger, select your car, manage users and "Add user". This gives you a QR code. You'll need to either use a second phoen or save this image on your computer or something so you can use it later.
2. Log in with your new account and scan the QR code.

**Comment by Amokd0c** [comment]
@dennzs ah ... OK, thanks very much! Your way was much much easier than mine! :)

**Comment by silashansen** [comment]
I really want to thank @ReneNulschDE for fixing this issue quickly, with the best solution possible, given the circumstances of their sudden rate-limiting policies <3
